### DISCUSSION TITLE
Samples for Seeed Studio XIAO nRF52840

### DISCUSSION THREAD

**Discussion by ChumbaWambalala** [ysoldak/HeadTracker#21] [General]
Hi @ysoldak , I've been looking for a headtracking solution and I stumbled on your project. It's nice to see that you are using alternative boards other than the Nano 33. I find it easier to source the Seed Studio Xiao compared to the Arduino. Was wondering if you could share how you setup the Xiao ? Looking forward to it.

**Comment by ysoldak** [comment]
Hi @ChumbaWambalala, glad you like this hobby project of mine. What do you mean "setup"?
You just plug XIAO BLE to your computer via USB, double-tap on RST button, a new mass storage device (like a USB stick or SD card) appears and you copy over respective `ht_xiao-ble_xxx.uf2` file to this device. XIAO BLE then reboots and you have your Head Tracker.

The binary .uf2 file you download here https://github.com/ysoldak/HeadTracker/releases

Pinout for XIAO board you can see, for example, [here](https://github.com/ysoldak/HeadTracker/blob/master/doc/xiao-ble-pinout.png), in case you want to wire an orientation reset button or use PPM output.

Give it a try and don't hesitate to ask any questions.
In case you like it, consider sponsoring the project :D This tells me the project is useful and incentivise adding new cool features to it.

**Reply by ChumbaWambalala** [reply]
Hi @ysoldak thank you for the reply. I think I've missed that you already mentioned on the docs bellow on how to setup the trainer mode via audio cable sorry.  Now that you've attached the pinouts reference this will be easier to understand. I'm really keen to test out the setup. Do you happen to have a debugging tool on the PC to confirm that the board is indeed giving orientation data?

**Reply by ysoldak** [reply]
If you happen to have Arduino IDE installed, you can use `Serial Console` and see printouts coming from the board when connected to PC with USB cable. Any other serial console shall do. Sorry, don't know any on PC but Arduino's. On Mac I use `picocom`.

Since you seem to plan to use it wired (PPM), does your radio support SBUS input and you'd want to use it instead? SBUS, being digital protocol, delivers clean signal, PPM is a bit noisy, to be honest. We have a [ticket](https://github.com/ysoldak/HeadTracker/issues/18) about adding support for SBUS out, just  trying gauge the interest and need.

I myself connect my HT to my radio via Bluetooth.

**Comment by jarTAT** [comment]
hi,@ysoldak ,I‘ve been rebuild the flash uf2 file. 
I try "tinygo build -target=xiao-ble main.go",
but get err msg "trainer\para.go:11:2: tinygo.org/x/bluetooth@v0.5.0: replacement directory ../bluetooth does not exist"

I already try "go mod tidy" and "git clone https://github.com/ysoldak/bluetooth.git "

How can I fix the project?
Where can I find the local files "../blutooth"

Thanks!

**Reply by ysoldak** [reply]
@jarTAT to put things straight first:
you don’t need rebuild the project just to flash your XIAO BLE board, you can find ready built UF2 file on releases page.

https://github.com/ysoldak/HeadTracker/releases

Assuming you really want to build you own UF2 file, probably with some modifications, you shall git clone https://github.com/ysoldak/bluetooth.git to a folder on the same level as HeadTracker (i.e. HeadTracker and bluetooth folders have same parent folder). Next, you have to switch to specific branch of bluetooth repository: `git switch sd-gatts-sys-attr` and then run `make build` in HeadTracker folder or look for exact build command in Makefile.

There is also an option instead of cloning bluetooth repository, you can simply replace “../bluetooth” with “ github.com/ysoldak/bluetooth sd-gatts-sys-attr” in go.mod file.
Hope that helps!

**Reply by jarTAT** [reply]
@ysoldak Thank you for the reply.
My RC does not support BT mode.
To get rid off the wires, I'm trying to add a smail BT module and set up the bridge between the xia sense and the RC trainer port.

**Reply by ysoldak** [reply]
@jarTAT in such case you may be interested in looking at this repository https://github.com/dlktdr/HeadTracker

They have a “helper” board to do just that. They use Nano 33 BLE both on goggles and on radio. These two boards communicate wirelessly. The board on radio side connects to radio with PPM cable.

Since the protocol is the same, OpenTTX wireless trainer, you can even use their helper board to talk to this, my, Bluetooth head tracker.

**Reply by ysoldak** [reply]
@jarTAT I've now updated `go.mod` file, so it shall just work w/o any extra dances.
You shall use TinyGo v0.27.0 and go 1.19
